src/bin/ion/commands/eq.rs

```
@@ -0,0 +1,278 @@
+use crate::ansi_codes::*;
+use crate::commands::eq::InputType::{Auto, File, Hex, Ion};
+use crate::commands::IonCliCommand;
+use crate::hex_reader::HexReader;
+use crate::input::CommandInput;
+use anyhow::Result;
+use clap::error::ErrorKind;
+use clap::{Arg, ArgAction, ArgMatches, Command, Error};
+use ion_rs::*;
+use std::fs;
+use std::io;
+use std::io::{Cursor, Read};
+use std::str::from_utf8;
+use std::sync::LazyLock;
+
+pub struct EqCommand;
+
+static HELP_EPILOGUE: LazyLock<String> = LazyLock::new(|| {
+    format!(
+        // '\' at the end of the line indicates that CLAP will handle the line wrapping.
+        "\
+Exactly two Ion streams must be provided, and may be provided as a file name, a string of \
+hexadecimal pairs, or a string of Ion text or binary. If only one stream is provided as an \
+argument, then stdin is implied to be the second Ion stream.
+
+Input mode can be specified up to two times. If input modes are provided, the first one applies \
+to the first input and the second one (if present) applies to the second input. No more than two \
+input mode flags may be provided. The input mode is not required to be specified. The eq command \
+will attempt to infer the type of each input using the following heuristic:
+    1. If the input is not valid UTF-8, it is assumed to be Ion binary
+    2. If the input is a path to an existing file, the input is assumed to be a file name
+    3. Then, eq attempts to parse the input as Ion
+    4. Then, eq attempts to parse the input as a stream of hex digit pairs
```

(I'm confident you've already thought through these cases, but I'll spell them out for other reviewers to mull.)

This is clever, but I think I'd prefer just explicitly flagging hex pairs or strings of Ion text. The inference would usually work, but there are weird corner cases that I'd rather not have to think about. For example: 
* If I try to read the file `example_file` and it's missing, the tool will consider the name to be the Ion symbol `example_file`. I think it would be better to surface an error when the file is missing.
* Many hex pair sequences that don't start with a binary IVM are legal text Ion streams. Consider `63 75 72 73 65 73 21`, which is both the ascii sequence `curses!` and a text Ion 1.0 integer sequence.

For consistency/predictability, I'd like unadorned arguments to always be considered file names.

---

> Consider 63 75 72 73 65 73 21, which is both the ascii sequence curses! and a text Ion 1.0 integer sequence

...but `curses!` is not a valid Ion text stream, so there's no ambiguity.

I've explained in more detail [here](https://github.com/amazon-ion/ion-cli/blob/aff324880d03c7a1570ec8ebf3f4b00d96b6f7fe/src/bin/ion/commands/eq.rs#L207-L218) why I think that it's okay, but I'm fine with removing the automatic detection if the consensus is that it is unhelpful. (Although, would you consider having a simpler file-vs-Ion auto-detection that a user can set as their own default by setting an environment variable?)

---

> ...but curses! is not a valid Ion text stream, so there's no ambiguity.

That's what I get for trying to be funny. 😅 Lop off the `!` and the point still stands.

I did see the other explanation later in my review. I think the cost/value of automatic detection is pretty neutral--it saves a small amount of typing and introduces some small corner cases. Personally I'd like `--text "my Ion string"` and `--hex 63 75 72 73 65 73`.

> (Although, would you consider having a simpler file-vs-Ion auto-detection that a user can set as their own default by setting an environment variable?)

Hm, maybe? I guess I'd have to see the design.